File: khex_com/khex_driver_base/samples/MATLAB/testSendQuadCmd1.m
% Copyright KMel Robotics 2014. Must read KMEL_LICENSE.pdf for terms and conditions before use.

function testSendQuadCmd1()

addpath ../../api
%open the interface
KQUAD.dev    = '/dev/ttyUSB0';
KQUAD.driver = @kQuadInterfaceAPI;
KQUAD.baud   = 921600;
KQUAD.id     = 7;
KQUAD.chan   = 1;
KQUAD.type   = 0; %0 for standard, 1 for nano

KQUAD.driver('connect',KQUAD.dev,KQUAD.baud);


while(1)
  thrust = 1; %grams. 1 for idle
  roll   = 0; %radians
  pitch  = 0; %radians
  yaw    = 0; %radians
  
  trpy = [thrust roll pitch yaw];
  
  KQUAD.driver('SendQuadCmd1',KQUAD.id, KQUAD.chan, KQUAD.type, trpy);

  pause(0.02);
  
  ch = getch();
    if ~isempty(ch)
        if ch == 'q'
            clear all
            break;
        end
    end
end

File: khex_com/khex_driver_base/samples/C++/testSendQuadCmd1.cpp
/* 
 * testSendQuadCmd1
 * 
 * Demonstrates sending control command using kQuadInterface.
 * SendQuadCmd1 sends desired thrust, roll, pitch, yaw.
 * 
 * Copyright KMel Robotics 2014. 
 * Must read KMEL_LICENSE.pdf for terms and conditions before use. 
*/

#include "kQuadInterface.hh"

void printUsage()
{
  std::cout << "Usage: ./testSendQuadCmd1 [PATH/TO/USB/DEVICE]" << std::endl;
}

int main(int argc, char* argv[])
{
  if (argc < 2)
  {
    printUsage();
    return -1;
  }
  
  char* dev = argv[1];
  int baud   = 921600;
  
  // kQuadInterface is the interface used to send/receive messages
  kQuadInterface kqi;
  
  // Connect to the device
  if (kqi.Connect(dev,baud))
  {
    printf("could not connect to the device\n");
    return -1;
  }
  
  // Initialize the send thread
  if (kqi.StartSendThread())
  {
    printf("could not start send thread\n");
    return -1;
  }
  
  // Initialize the receiving thread; won't be used in this example
  if (kqi.StartRecvThread())
  {
    printf("could not start receive thread\n");
    return -1;
  }
  
  uint8_t quadId   = 7; //must match the vehicle id
  uint8_t quadType = 0; //0 for standard and Nano+, 1 for Nano only
  uint8_t channel  = 1; //zigbee channel, does not matter for uart connection
  
  float thrust = 1; //(grams) 1 for idle
  float roll   = 0; //radians
  float pitch  = 0; //radians
  float yaw    = 0; //radians
  
  list<QuadStatusData> qdata;
  
  while(1)
  {
    kqi.SendQuadCmd1(quadId, quadType, channel, thrust, roll, pitch, yaw);
    // int nmsg  = kqi.GetQuadStatusData(qdata);
    
    auto p = qdata.end(); p--;
    
    std::cout << p->state << std::endl;
    
    usleep(20000);
    
    //printf("."); fflush(stdout);
  }
  
 
  return 0;
}
